File: companies/Zoho/120.triangle.java
import java.util.*;

/**
 * Algorithm:
 * Bottom-up DP: the best path from a cell is its value plus the cheaper of the
 * two adjacent best paths below it. Initialize dp with the last row and update
 * upward in place.
 *
 * Java data structures:
 * int[] dp stores one row of dynamic programming values.
 *
 * Complexity:
 * Time O(total cells), space O(number of rows).
 */
class Solution {
    public int minimumTotal(List<List<Integer>> triangle) {
        int n = triangle.size();
        int[] dp = new int[n];
        for (int i = 0; i < n; i++) {
            dp[i] = triangle.get(n - 1).get(i);
        }
        for (int r = n - 2; r >= 0; r--) {
            for (int c = 0; c < triangle.get(r).size(); c++) {
                dp[c] = triangle.get(r).get(c) + Math.min(dp[c], dp[c + 1]);
            }
        }
        return dp[0];
    }
}

File: Microsoft/120.triangle.cpp
#include <algorithm>
#include <cctype>
#include <climits>
#include <cmath>
#include <cstdlib>
#include <functional>
#include <numeric>
#include <queue>
#include <sstream>
#include <string>
#include <unordered_map>
#include <utility>
#include <vector>
using namespace std;


class Solution {
public:
    int minimumTotal(vector<vector<int>>& triangle) {
        /*
        Approach:
        Bottom-up DP: the best path from a cell equals its value plus the
        cheaper best path from the two adjacent cells below. Initialize dp to the
        last row and update upward in place.

        C++ notes:
        vector<int> dp copies the last row and provides O(rows) extra storage.

        Complexity: O(total cells) time and O(number of rows) space.
        */
        vector<int> dp = triangle.back();
        for (int r = (int)triangle.size() - 2; r >= 0; --r) {
            for (int c = 0; c < (int)triangle[r].size(); ++c) {
                dp[c] = triangle[r][c] + min(dp[c], dp[c + 1]);
            }
        }
        return dp[0];
    }
};
